### PR TITLE
2017-06-14-Ansible-via-python.markdown

### PR DESCRIPTION
**Pull request by Maciej-Rucinski-lofi**

### CODE REVIEW

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 1]
```diff
@@ -0,0 +1,188 @@
+# Ansible via Python ![alt text](https://cub.nobleprog.com/sites/hitramx/files/styles/height50_scale/public/category_image/cursos-de-ansible-en-mexico.png?itok=xPUrGNrA)
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown:10]: @Macias118 Please add header with author name, category etc. Example https://github.com/3mdeb/blog.3mdeb.com/pull/7/files

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+## What is Ansible
+
+Ansible is simple IT engine for automation, it is designed for manage many
+systems, rather than just one at a time. Ansible automates cloud provisioning,
+configuration management, application deployment, intra-service
+orchestration, and many other IT operations. It is easy to deploy due to using
+no agents and no additional custom security infrastructure. We can define own
+configuration management in simple YAML language, which is named
+ansible-playbook. YAML is easier to read and write by humans than other common
+data formats line XML or JSON. Futhermore, most programming languages contain
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 like

File: source/_posts/2017-06-14-Ansible-via-python.markdown (hunk 1 above)
Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown:10]: @Macias118 missing header, please take a look at other blog posts.

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+[dbservers]
+one.example.com
+two.example.com
+three.example.com
+```
+
+## Ansible Playbook
+
+Playbook is the way to direct systems, which is particularly powerful, compact
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown:48]: @Macias118 not sure if `direct` is correct in this context. Did you mean to manage systems?

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+You also need to have installed python, of course.
+
+#### Additional information
+
+For more information go to [Ansible Documentation](http://docs.ansible.com/ansible/)
+
+## Python API
+
+Python API is very powerful, so we can manage and run ansible-playbook drom
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 from

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+python level, there is possibility to controll nodes, write various plugins,
+extend Ansible to respond to various python events and plug in inventory data
+from external data sources.
+
+> Note: There is a permament structure to build python program which operates ansible
+commands:
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 another `>` missing?

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+First of all we have to import some modules needed to run ansible in python.
+
+- Let's describe some of them:
+
+	- `json` module to convert output to json format
+	- `ansible` modules to manage e.g. inventory or plays
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 module

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+      service: name=httpd state=restarted
+```
+
+### Ansible for Embedded Linux
+
+There is possibility to need build image with custom Linux-based system for
+embedded with Ansible. In addition, we would like to run ansible-playbook via
+python. It seems to be hard to implement. Nothing more simple! It is needed to
+add recipe to image with ansible from
+[Python Ansible package](https://github.com/OverC/meta-overc/blob/master/meta-cube/recipes-devtools/python/python-ansible_2.1.1.0.bb)
```
Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 please add some context. We cannot assume that Embedded Linux equals Yocto. Please be clear that you wrote about Yocto meta layer with support for Ansible.

File: source/_posts/2017-06-14-Ansible-via-python.markdown (hunk 4 above)
Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 dorm ?

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+```python
+class ResultCallback(CallbackBase):
+
+	def v2_runner_on_ok(self, result, **kwargs):
+		host = result._host
+		print json.dumps({host.name: result._result}, indent=4)
+```
+
+We can overrides more method. All specification can find in
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 ovverride; can be found

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+finally:
+    if tqm is not None:
+        tqm.cleanup()
+```
+
+## Conclusion
+
+Ansible delivers IT automation that ends repetitive tasks and frees up DevOps
+teams for more strategic work. Manage the Ansbile via Python API is easy, it
+can by apply to operate a configuration on many systems at the time, using
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 can be applied

Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 can be

**Review by pietrushnic — CHANGES_REQUESTED** [review]

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+orchestration, and many other IT operations. It is easy to deploy due to using
+no agents and no additional custom security infrastructure. We can define own
+configuration management in simple YAML language, which is named
+ansible-playbook. YAML is easier to read and write by humans than other common
+data formats line XML or JSON. Futhermore, most programming languages contain
+libraries which operate and work YAML.
+
+## Inventory
+
+Ansible works collaterally on many systems in your infrastructure, so it is
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown:28]: @Macias118 I find it difficult to understand the meaning of this sentence

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+  - name: ensure apache is running (and enable it at boot)
+    service: name=httpd state=started enabled=yes
+  handlers:
+    - name: restart apache
+      service: name=httpd state=restarted
+```
+
+### Ansible for Embedded Linux
+
+There is possibility to need build image with custom Linux-based system for
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 add note that this paragraph is relevant for Yocto build system

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+embedded with Ansible. In addition, we would like to run ansible-playbook via
+python. It seems to be hard to implement. Nothing more simple! It is needed to
+add recipe to image with ansible from
+[Python Ansible package](https://github.com/OverC/meta-overc/blob/master/meta-cube/recipes-devtools/python/python-ansible_2.1.1.0.bb)
+
+You also need to have installed python, of course.
```
Comment (macpijan) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 Actually, you don't due to recipes dependencies. So this sentence is not needed

**Review by macpijan — CHANGES_REQUESTED** [review]

File: source/_posts/2017-06-14-Ansible-via-python.markdown (hunk 10 above)
Comment (Maciej-Rucinski-lofi) [reply source/_posts/2017-06-14-Ansible-via-python.markdown:28]: @macpijan This means, that Ansible services systems in paralell way, using only one set of tasks.

File: source/_posts/2017-06-14-Ansible-via-python.markdown (hunk 3 above)
Comment (Maciej-Rucinski-lofi) [reply source/_posts/2017-06-14-Ansible-via-python.markdown:48]: @macpijan Of course, but i had to find synonym to 'manage', cause this word is used too often.

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 13]
```diff
@@ -1,3 +1,12 @@
+---
+author: Maciej Ruciński
+layout: post
+title: "Ansible via Python"
+date: 2017-06-14 12:00:00 +0100
+comments: true
+categories: Development
```
Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 I would say `python ansible`

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+```python
+class ResultCallback(CallbackBase):
+
+  def v2_runner_on_ok(self, result, **kwargs):
+    host = result._host
+    print json.dumps({host.name: result._result}, indent=4)
+```
+
+We can override more method. All specification can be found in
```
Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown]: @Macias118 metods

File: source/_posts/2017-06-14-Ansible-via-python.markdown [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+from collections import namedtuple
+from ansible.parsing.dataloader import DataLoader
+from ansible.vars import VariableManager
+from ansible.inventory import Inventory
+from ansible.playbook.play import Play
+from ansible.executor.task_queue_manager import TaskQueueManager
+from ansible.plugins.callback import CallbackBase
+```
+
+- ResultCallback which inherits from `CallbackBase` and manage the output of
```
Comment (pietrushnic) [inline source/_posts/2017-06-14-Ansible-via-python.markdown:122]: @Macias118 please for this and other points add why we need to do that. Description of each step is to mechanical. We don't want to have copy&paste-kind post we want to have understand-kind of post.

**Review by pietrushnic — CHANGES_REQUESTED** [review]

File: source/_posts/2017-06-14-Ansible-via-python.markdown (hunk 15 above)
Comment (pietrushnic) [reply source/_posts/2017-06-14-Ansible-via-python.markdown:122]: @Macias118 I don't feel you changed a lot in recent commit. Am I missing something ? Please elaborate on each line with at least 2 sentences. Please give reader feeling that you understand this code.

**Review by pietrushnic — CHANGES_REQUESTED** [review]

**Comment by pietrushnic** [comment]
@Macias118 please take a look at my comments.

**Comment by pietrushnic** [comment]
@Macias118 please address comments.

**Review by pietrushnic — APPROVED** [review]

**Comment by pietrushnic** [comment]
@macpijan @mek-x please review or/and approve.

**Review by macpijan — APPROVED** [review]

**Pull request merged**